File: definitions/loaders/VarbitLoader.cs
/*
 * Copyright (c) 2017, Adam <Adam@sigterm.info>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
namespace OSRSCache.definitions.loaders
{
	using VarbitDefinition = OSRSCache.definitions.VarbitDefinition;
	using InputStream = OSRSCache.io.InputStream;

	public class VarbitLoader
	{
		public virtual VarbitDefinition load(int id, byte[] b)
		{
			VarbitDefinition def = new VarbitDefinition(id);
			InputStream @is = new InputStream(b);

			for (;;)
			{
				int opcode = @is.readUnsignedByte();
				if (opcode == 0)
				{
					break;
				}

				if (opcode == 1)
				{
					def.index = @is.readUnsignedShort();
					def.leastSignificantBit = @is.readUnsignedByte();
					def.mostSignificantBit = @is.readUnsignedByte();
				}
			}

			return def;
		}
	}

}

File: src/main/java/net/runelite/cache/definitions/loaders/VarbitLoader.cpp
#include "VarbitLoader.h"

namespace net::runelite::cache::definitions::loaders
{
	using VarbitDefinition = net::runelite::cache::definitions::VarbitDefinition;
	using InputStream = net::runelite::cache::io::InputStream;

	std::shared_ptr<VarbitDefinition> VarbitLoader::load(int id, std::vector<signed char> &b)
	{
		std::shared_ptr<VarbitDefinition> def = std::make_shared<VarbitDefinition>();
		std::shared_ptr<InputStream> is = std::make_shared<InputStream>(b);

		def->setId(id);

		for (;;)
		{
			int opcode = is->readUnsignedByte();
			if (opcode == 0)
			{
				break;
			}

			if (opcode == 1)
			{
				def->setIndex(is->readUnsignedShort());
				def->setLeastSignificantBit(is->readUnsignedByte());
				def->setMostSignificantBit(is->readUnsignedByte());
			}
		}

		return def;
	}
}
